src/process/index/Indexer.h

```diff
@@ -0,0 +1,202 @@
+
+    path storageForPartialDecompressedBlocks;
+
+    ofstream partialBlockinfoStream;
+
+    /**
+     * Current bits for the next index entry.
+     */
+    int curBits{0};
+
+    /**
+     * Current offset in file.
+     */
+    long offset{0};
+
+    /**
+     * Marks, if the last inflated block ended with the newline character.
+     * If true, the blockOffsetInRawFile for the first line in the new block will be 0.
+     * If false, we have to look for the byte blockOffsetInRawFile.
+     */
+    bool lastBlockEndedWithNewline = true;
+
+    int64_t lineCountForNextIndexEntry{0};
+
+    int64_t numberOfConcatenatedFiles{1};
+
+    long blockID{-1};                   // Number of the currently processed block.
+
+    shared_ptr<IndexEntryV1> lastStoredEntry;
+
+    /**
+     * When we write out a new index entry, we need a dictionary. This is taken from the block of data before the block
+     * to which the new entry references.
+     */
+    Bytef dictionaryForNextBlock[WINDOW_SIZE]{0};
+
+public:
+
+
+    /**
+     * Be careful, when you enableDebugging. This will tell the Indexer to store information about the process, which can
+     * e.g. be used for unit test. E.g. this will store ALL lines found in the FASTQ file! So it is absolutely not
+     * advisable to use it with large FASTQ files. For test data it is safe to use.
+     * @param fastq The FASTQ we are working on.
+     * @param index The index for the FASTQ.
+     * @param enableDebugging Store debug information or not.
+     */
+    Indexer(const shared_ptr<Source> &fastqfile,
+            const shared_ptr<Sink> &index,
+            const shared_ptr<IndexEntryStorageStrategy> &storageStrategy,
+            bool enableDebugging = false,
+            bool forceOverwrite = false,
+            bool forbidWriteFQI = false,
+            bool compressDictionaries = true
+    );
+
+    ~Indexer() override = default;
+
+    void setDictionaryCompression(bool value) {
+        this->compressDictionaries = value;
+    }
+
+    bool fulfillsPremises();
+
+    /**
+     * This will create an IndexHeader instance with INDEXER_VERSION and the size of the used IndexEntry struct.
+     */
+    shared_ptr<IndexHeader> createHeader();
+
+    /**
+     * Start the index creation,
+     * @return true, if everything went fine.
+     */
+    bool createIndex();
+
+    void finalizeProcessingForCurrentBlock(stringstream &currentDecompressedBlock, z_stream *strm);
+
+    bool checkAndPrepareForNextConcatenatedPart();
+
+    void storeLinesOfCurrentBlockForDebugMode(std::stringstream &currentDecompressedBlock);
+
+    /**
+     * Overriden to also pass through (copywise, safe but slow but also only with a few entries and in error cases)
+     * error messages from the used IndexWriter and ZLibHelper instance.
+     * @return Merged vector of the objects error messages + the index writers error messages.
+     */
+    vector<string> getErrorMessages() override;
```

Comment:
Did you consider making a dedicated Error return value class? For accumulating errors? You could add methods for easy accumulation and merging.